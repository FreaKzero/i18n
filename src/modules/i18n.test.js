import useTranslation, { setupTranslations } from "./i18n";

import React from "react";
import { shallow } from "enzyme";

setupTranslations({
  multi: "There are {{count}} {{animal}}",
  nonamespace: "This is a message without namespace",
  mynamespace: {
    test: "This is a {{test}}",
  },
  multimix: {
    mm: "There is {{b}} {{a}}",

    mm_plural: "There are {{b}} {{a}}",
  },
  plmix: {
    doing: "There is {{count}} cats {{doing}} (default)",
    doing_zero: "There are {{count}} cats {{doing}} (zero)",
    doing_one: "There is {{count}} cat {{doing}} (one)",
    doing_plural: "There are {{count}} cats {{doing}} (plural)",
  },
  plall: {
    cat: "There is {{count}} cat (default)",
    cat_zero: "There are {{count}} cats (zero)",
    cat_one: "There is {{count}} cat (one)",
    cat_plural: "There are {{count}} cats (plural)",
  },
  pldefzero: {
    dog: "There is {{count}} dog (default)",
    dog_plural: "There are {{count}} dogs (plural)",
  },
  plundefined: {
    dog_plural: "There are {{count}} dogs",
  },
  plzeroone: {
    cat: "There are {{count}} cats (default)",
    cat_zero: "There should be {{count}} cats (zero)",
    cat_one: "There should be {{count}} cat (one)",
  },
});

describe("useTranslation Hook", () => {
  const Test = ({ message, replace, namespace }) => {
    const { t } = useTranslation(namespace);
    return <h1>{t(message || "test", replace)}</h1>;
  };

  const Multitest = () => {
    const { t } = useTranslation();
    return <h1>{t("multi", { count: 3, animal: "cats" })}</h1>;
  };

  const PluralTest = ({ namespace, message }) => {
    const { t } = useTranslation(namespace);
    return (
      <ul>
        {[0, 1, 2, 5].map((count) => (
          <li key={`item_${count}`}>{t(message, { count })}</li>
        ))}
      </ul>
    );
  };

  const MixPluralTest = ({ namespace, message }) => {
    const { t } = useTranslation(namespace);
    return (
      <ul>
        {[0, 1, 2, 5].map((count) => (
          <li key={`item_${count}`}>
            {t(message, { count, doing: "jumping" })}
          </li>
        ))}
      </ul>
    );
  };

  const PluralMultiMix = ({ namespace, message }) => {
    const { t } = useTranslation(namespace);
    return (
      <ul>
        {[0, 1, 2, 5].map((count) => (
          <li key={`item_${count}`}>{t(message, { a: "cats", b: count })}</li>
        ))}
      </ul>
    );
  };

  it("replacement order", () => {
    const component = shallow(
      <PluralMultiMix namespace="multimix" message="mm" />
    );
    expect(component).toMatchSnapshot();
  });

  it("plurals work with mixed replacements", () => {
    const component = shallow(
      <MixPluralTest namespace="plmix" message="doing" />
    );
    expect(component).toMatchSnapshot();
  });

  it("works with default, _zero, _one", () => {
    const component = shallow(
      <PluralTest namespace="plzeroone" message="cat" />
    );
    expect(component).toMatchSnapshot();
  });

  it("Shows translationstring when replacevalues are not string or number ", () => {
    const component = shallow(
      <Test namespace="mynamespace" replace={{ test: { a: "b" } }} />
    );
    expect(component).toMatchSnapshot();
  });

  it("cant find key when only plurals are defined", () => {
    const component = shallow(
      <PluralTest namespace="plundefined" message="dog" />
    );
    expect(component).toMatchSnapshot();
  });

  it("Plurals default and _plural", () => {
    const component = shallow(
      <PluralTest namespace="pldefzero" message="dog" />
    );
    expect(component).toMatchSnapshot();
  });

  it("Plurals _zero, _one and _plural", () => {
    const component = shallow(<PluralTest namespace="plall" message="cat" />);
    expect(component).toMatchSnapshot();
  });

  it("Replaces strings with tpl parameter", () => {
    const component = shallow(
      <Test namespace="mynamespace" replace={{ test: "replaced string" }} />
    );
    expect(component).toMatchSnapshot();
  });

  it("Replaces numbers with tpl parameter", () => {
    const component = shallow(
      <Test namespace="mynamespace" replace={{ test: 1 }} />
    );
    expect(component).toMatchSnapshot();
  });

  it("Shows the translationstring when not found", () => {
    const component = shallow(
      <Test namespace="mynamespace" message="nokey" replace={{ test: 1 }} />
    );
    expect(component).toMatchSnapshot();
  });

  it("Shows the translationstring without namespace when not found", () => {
    const component = shallow(<Test message="nokey" replace={{ test: 1 }} />);
    expect(component).toMatchSnapshot();
  });

  it("Translates without namespace and without template", () => {
    const component = shallow(<Test message="nonamespace" />);
    expect(component).toMatchSnapshot();
  });

  it("replaces multiple templates", () => {
    const component = shallow(<Multitest />);
    expect(component).toMatchSnapshot();
  });

  it("does not crash when namespace is not defined", () => {
    const component = shallow(<Test namespace="notdefined" message="nokey" />);
    expect(component).toMatchSnapshot();
  });
});
