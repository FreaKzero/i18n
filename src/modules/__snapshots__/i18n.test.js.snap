// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTranslation Hook Plurals _zero, _one and _plural 1`] = `
<ul>
  <li
    key="item_0"
  >
    There are 0 cats (zero)
  </li>
  <li
    key="item_1"
  >
    There is 1 cat (one)
  </li>
  <li
    key="item_2"
  >
    There are 2 cats (plural)
  </li>
  <li
    key="item_5"
  >
    There are 5 cats (plural)
  </li>
</ul>
`;

exports[`useTranslation Hook Plurals default and _plural 1`] = `
<ul>
  <li
    key="item_0"
  >
    There are 0 dogs (plural)
  </li>
  <li
    key="item_1"
  >
    There is 1 dog (default)
  </li>
  <li
    key="item_2"
  >
    There are 2 dogs (plural)
  </li>
  <li
    key="item_5"
  >
    There are 5 dogs (plural)
  </li>
</ul>
`;

exports[`useTranslation Hook Replaces numbers with tpl parameter 1`] = `
<h1>
  This is a 1
</h1>
`;

exports[`useTranslation Hook Replaces strings with tpl parameter 1`] = `
<h1>
  This is a replaced string
</h1>
`;

exports[`useTranslation Hook Shows the translationstring when not found 1`] = `
<h1>
  {{ MYNAMESPACE-NOKEY }}
</h1>
`;

exports[`useTranslation Hook Shows the translationstring without namespace when not found 1`] = `
<h1>
  {{ NOKEY }}
</h1>
`;

exports[`useTranslation Hook Shows translationstring when replacevalues are not string or number  1`] = `
<h1>
  This is a {{ MYNAMESPACE-TEST }}
</h1>
`;

exports[`useTranslation Hook Translates without namespace and without template 1`] = `
<h1>
  This is a message without namespace
</h1>
`;

exports[`useTranslation Hook cant find key when only plurals are defined 1`] = `
<ul>
  <li
    key="item_0"
  >
    {{ PLUNDEFINED-DOG }}
  </li>
  <li
    key="item_1"
  >
    {{ PLUNDEFINED-DOG }}
  </li>
  <li
    key="item_2"
  >
    {{ PLUNDEFINED-DOG }}
  </li>
  <li
    key="item_5"
  >
    {{ PLUNDEFINED-DOG }}
  </li>
</ul>
`;

exports[`useTranslation Hook does not crash when namespace is not defined 1`] = `
<h1>
  {{ NOTDEFINED-NOKEY }}
</h1>
`;

exports[`useTranslation Hook plurals work with mixed replacements 1`] = `
<ul>
  <li
    key="item_0"
  >
    There are 0 cats jumping (zero)
  </li>
  <li
    key="item_1"
  >
    There is 1 cat jumping (one)
  </li>
  <li
    key="item_2"
  >
    There are 2 cats jumping (plural)
  </li>
  <li
    key="item_5"
  >
    There are 5 cats jumping (plural)
  </li>
</ul>
`;

exports[`useTranslation Hook replacement order 1`] = `
<ul>
  <li
    key="item_0"
  >
    There are 0 cats
  </li>
  <li
    key="item_1"
  >
    There is 1 cats
  </li>
  <li
    key="item_2"
  >
    There are 2 cats
  </li>
  <li
    key="item_5"
  >
    There are 5 cats
  </li>
</ul>
`;

exports[`useTranslation Hook replaces multiple templates 1`] = `
<h1>
  There are 3 cats
</h1>
`;

exports[`useTranslation Hook works with default, _zero, _one 1`] = `
<ul>
  <li
    key="item_0"
  >
    There should be 0 cats (zero)
  </li>
  <li
    key="item_1"
  >
    There should be 1 cat (one)
  </li>
  <li
    key="item_2"
  >
    There are 2 cats (default)
  </li>
  <li
    key="item_5"
  >
    There are 5 cats (default)
  </li>
</ul>
`;
